src/rotator_library/dynamic_chain.py

```
@@ -0,0 +1,365 @@
+            conn.row_factory = sqlite3.Row
+            # Pull only window rows for the candidate providers. Aggregation
+            # (EMA, last-failure) is done in Python because SQLite has no EMA.
+            qmarks = ",".join("?" * len(candidates))
+            rows = conn.execute(
+                f"""
+                SELECT provider, ts_start, status
+                FROM llm_events
+                WHERE ts_start >= ?
+                  AND lower(provider) IN ({qmarks})
+                ORDER BY ts_start ASC
+                """,
```

Ruff (0.15.18)

[error] Possible SQL injection vector through string-based query construction

(S608)